draft-ietf-tls-tls13.md

```diff
@@ -4319,6 +4322,31 @@ will receive various TLS 1.x versions in this field, however its value
 MUST always be ignored.
 
 
+## Zero-RTT backwards compatibility
+
+A TLS 1.3 ClientHello with 0-RTT data is not compatible with older servers
+implementing previous versions of TLS. This can cause compatibility issues if a
+client first connects to a 0-RTT-capable TLS 1.3 server and later to an older
+server, both behind the same service. For example, a multi-server deployment
+may deploy TLS 1.3 gradually with some servers implementing TLS 1.3 and some
+implementing TLS 1.2. This may also occur if a TLS 1.3 deployment gets reverted
+to TLS 1.2.
+
+If a TLS 1.3 client sends a ClientHello with 0-RTT to an older server, the
+server will respond with a ServerHello containing an older version number.
+(The handshake will later fail at the 0-RTT data.) Clients which accept older
+versions of TLS SHOULD then retry the connection advertising TLS 1.3 without
+offering 0-RTT.
```

Comment:
Not the same one as before. Rather than falling back on generic error and insecurely dropping the version, this is just a retry on a particular signal and punting on 0-RTT. The ClientHello version should be unchanged. (Though the old fallback is probably going to come back too. I probed for 1.3 intolerance way back and it's everywhere. :-( )

Agreed that this is incredibly nasty. I can't think of a better solution though, assuming we want to continue allowing streaming 0-RTT. I think it's unrealistic to expect that servers will get the deployment considerations correct. (I'm thinking people who just install their favorite HTTP server from their favorite OS distribution and put a few of them behind a hostname.)

Given that, I think it's better that we spell out the fallback we want clients to implement so they don't make worse mistakes (dropping ClientHello.version or using an overly broad fallback signal that masks future implementation bugs).